core/camera/src/main/java/com/google/jetpackcamera/core/camera/CameraSession.kt

```diff
@@ -99,63 +100,124 @@ internal suspend fun runSingleCameraSession(
-                useCaseGroup.getVideoCapture(),
-                captureTypeSuffix = when (sessionSettings.captureMode) {
-                    CaptureMode.MULTI_STREAM -> "MultiStream"
-                    CaptureMode.SINGLE_STREAM -> "SingleStream"
-                }
-            )
-        }
+        cameraProvider.runWith(
+            currentTransientSettings.cameraInfo.cameraSelector,
+            useCaseGroup
+        ) { camera ->
+            Log.d(TAG, "Camera session started")
+            // process torch while video is recording
+            launch {
+                // only collect when starting or stopping recording
+                currentCameraState
+                    .distinctUntilChanged { old, new ->
+                        (old.videoRecordingState is VideoRecordingState.Active) ==
+                            (new.videoRecordingState is VideoRecordingState.Active) ||
+                            (old.videoRecordingState is VideoRecordingState.Inactive) ==
+                            (new.videoRecordingState is VideoRecordingState.Inactive)
+                    }
+                    .collectLatest {
+                        // todo(): How should we handle torch on Auto FlashMode?
+                        if (it.videoRecordingState is VideoRecordingState.Active &&
+                            currentTransientSettings.flashMode == FlashMode.ON
+                        ) {
+                            toggleTorch(camera, true)
+                        } else {
+                            toggleTorch(camera, false)
+                        }
+                    }
+            }
 
-        launch {
-            camera.cameraInfo.torchState.asFlow().collectLatest { torchState ->
-                currentCameraState.update { old ->
-                    old.copy(torchEnabled = torchState == TorchState.ON)
+            launch {
+                processFocusMeteringEvents(camera.cameraControl)
+            }
+
+            launch {
+                camera.cameraInfo.torchState.asFlow().collectLatest { torchState ->
+                    currentCameraState.update { old ->
+                        old.copy(torchEnabled = torchState == TorchState.ON)
+                    }
                 }
             }
+
+            applyDeviceRotation(currentTransientSettings.deviceRotation, useCaseGroup)
+            processTransientSettingEvents(
+                camera,
+                useCaseGroup,
+                currentTransientSettings,
+                transientSettings
+            )
         }
+    }
+}
 
-        applyDeviceRotation(initialTransientSettings.deviceRotation, useCaseGroup)
-        processTransientSettingEvents(
-            camera,
-            useCaseGroup,
-            initialTransientSettings,
-            transientSettings
-        )
+context(CameraSessionContext)
+/**
+ * enables torch for the provided camera based on the current flash settings
+ */
+suspend fun toggleTorch(camera: Camera, newTorchOn: Boolean) {
```

Comment:
This isn't really toggling the torch as much as it's setting the torch. Maybe call it `setTorch` instead.